sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala

```diff
@@ -65,6 +65,50 @@ case class CollectLimitExec(limit: Int, child: SparkPlan) extends LimitExec {
   }
 }
 
+/**
+ * Take the last `limit` elements and collect them to a single partition.
+ *
+ * This operator will be used when a logical `Tail` operation is the final operator in an
+ * logical plan, which happens when the user is collecting results back to the driver.
+ */
+case class CollectTailExec(limit: Int, child: SparkPlan) extends LimitExec {
+  override def output: Seq[Attribute] = child.output
+  override def outputPartitioning: Partitioning = SinglePartition
+  override def executeCollect(): Array[InternalRow] = child.executeTail(limit)
+  private val serializer: Serializer = new UnsafeRowSerializer(child.output.size)
+  private lazy val writeMetrics =
+    SQLShuffleWriteMetricsReporter.createShuffleWriteMetrics(sparkContext)
+  private lazy val readMetrics =
+    SQLShuffleReadMetricsReporter.createShuffleReadMetrics(sparkContext)
+  override lazy val metrics = readMetrics ++ writeMetrics
+  protected override def doExecute(): RDD[InternalRow] = {
```

Comment:
looks like `executeTail` scans partitions from the end until returning last `n` rows? This `doExecute` looks like different to it?

Reply:
Yes, I noticed it too in `executeTake` vs `Limit`. It's a bit different.

`executeTail` and `doExecute` here do the same thing but in a different way. I think it's similar with `limit(...).collect()` vs  `limit(...).foreach()`.

For `tail` case, actually, we can even remove all such plans at all (because we will currently only have one API `tail(...)` alone unlike `limit` for now. This was a pain point when I worked on this PR. 

I just decided to be consistent with `Limit` considering the case we push down such plans into DataSource V2 later or we expose this plan as another API (or use it in another API). I can remove this if any of you feels strongly.